File: src/Core/Components/Grid/FluentGrid.razor.cs
// ------------------------------------------------------------------------
// MIT License - Copyright (c) Microsoft Corporation. All rights reserved.
// ------------------------------------------------------------------------

using System.Diagnostics.CodeAnalysis;
using Microsoft.AspNetCore.Components;
using Microsoft.FluentUI.AspNetCore.Components.Utilities;
using Microsoft.JSInterop;

namespace Microsoft.FluentUI.AspNetCore.Components;

/// <summary>
/// The grid component helps keeping layout consistent across various screen resolutions and sizes.
/// PowerGrid comes with a 12-point grid system and contains 5 types of breakpoints
/// that are used for specific screen sizes.
/// </summary>
public partial class FluentGrid : FluentComponentBase
{
    private const string JAVASCRIPT_FILE = FluentJSModule.JAVASCRIPT_ROOT + "Grid/FluentGrid.razor.js";

    /// <summary>
    /// Initializes a new instance of the <see cref="FluentGrid"/> class.
    /// </summary>
    public FluentGrid()
    {
        Id = Identifier.NewId();
    }

    /// <summary />
    protected string? ClassValue => DefaultClassBuilder
        .Build();

    /// <summary />
    protected string? StyleValue => DefaultStyleBuilder
        .Build();

    /// <summary />
    internal GridItemSize? CurrentSize { get; private set; }

    /// <summary>
    /// Gets or sets the distance between flexbox items, using a multiple of 4px.
    /// Only values from 0 to 10 are possible.
    /// </summary>
    [Parameter]
    public int Spacing { get; set; } = 3;

    /// <summary>
    /// Defines how the browser distributes space between and around content items.
    /// </summary>
    [Parameter]
    public JustifyContent Justify { get; set; } = JustifyContent.FlexStart;

    /// <summary>
    /// Gets or sets the adaptive rendering, which not render the HTML code when the item is hidden (true) or only hide the item by CSS (false).
    /// Default is false.
    /// </summary>
    [Parameter]
    public bool AdaptiveRendering { get; set; } = false;

    /// <summary>
    /// Gets or sets the child content of component.
    /// </summary>
    [Parameter]
    public RenderFragment? ChildContent { get; set; }

    /// <summary>
    /// Event raised when page size falls within a specific size range (xs, sm, md, lg, xl, xxl).
    /// </summary>
    [Parameter]
    public EventCallback<GridItemSize> OnBreakpointEnter { get; set; }

    /// <summary />
    protected override async Task OnAfterRenderAsync(bool firstRender)
    {
        if (firstRender && OnBreakpointEnter.HasDelegate)
        {
            // Import the JavaScript module
            var jsModule = await JSModule.ImportJavaScriptModuleAsync(JAVASCRIPT_FILE);

            // Call a function from the JavaScript module
            var dotNetHelper = DotNetObjectReference.Create(this);
            await jsModule.InvokeVoidAsync("Microsoft.FluentUI.Blazor.Grid.FluentGridInitialize", Id, dotNetHelper);
        }
    }

    /// <summary />
    [JSInvokable]
    public async Task FluentGrid_MediaChangedAsync(string size)
    {
        var valid = Enum.TryParse<GridItemSize>(size, ignoreCase: true, out var sizeEnum);
        CurrentSize = valid ? sizeEnum : null;

        if (OnBreakpointEnter.HasDelegate)
        {
            if (valid)
            {
                await OnBreakpointEnter.InvokeAsync(sizeEnum);
            }
        }
    }

    /// <summary>
    /// <inheritdoc cref="IAsyncDisposable.DisposeAsync" />
    /// </summary>
    /// <returns></returns>
    [ExcludeFromCodeCoverage(Justification = "Tested via integration tests.")]
    protected override async ValueTask DisposeAsync(IJSObjectReference? jsModule)
    {
        if (jsModule != null)
        {
            await jsModule.InvokeVoidAsync("Microsoft.FluentUI.Blazor.Grid.FluentGridCleanup", Id);
        }
    }
}

File: src/Core/Components/Grid/FluentGrid.razor.ts
export namespace Microsoft.FluentUI.Blazor.Grid {

  function GetMediaQueries() {
    return [
      { id: 'xs', items: (document as any)._fluentGrid.mediaXS, query: '(max-width: 599.98px)' },
      { id: 'sm', items: (document as any)._fluentGrid.mediaSM, query: '(min-width: 600px) and (max-width: 959.98px)' },
      { id: 'md', items: (document as any)._fluentGrid.mediaMD, query: '(min-width: 960px) and (max-width: 1279.98px)' },
      { id: 'lg', items: (document as any)._fluentGrid.mediaLG, query: '(min-width: 1280px) and (max-width: 1919.98px)' },
      { id: 'xl', items: (document as any)._fluentGrid.mediaXL, query: '(min-width: 1920px) and (max-width: 2559.98px)' },
      { id: 'xxl', items: (document as any)._fluentGrid.mediaXXL, query: '(min-width: 2560px)' },
    ];
  }

  export function FluentGridInitialize(id: string, dotNetHelper: any) {

    // Create a single instance of the media queries
    if (!(document as any)._fluentGrid) {
      (document as any)._fluentGrid = {
        mediaXS: [],
        mediaSM: [],
        mediaMD: [],
        mediaLG: [],
        mediaXL: [],
        mediaXXL: [],
      }

      // Add event listeners for each media query
      GetMediaQueries().forEach((mediaQuery) => {
        window.matchMedia(mediaQuery.query)
          .addEventListener('change', media => {
            if (media.matches) {
              mediaQuery.items.forEach((item: any) => {
                item.dotNetHelper.invokeMethodAsync('FluentGrid_MediaChangedAsync', mediaQuery.id);
              });
            }
          });
      });
    }

    // Add the item to each media query
    (document as any)._fluentGrid.mediaXS.push({ id: id, dotNetHelper: dotNetHelper });
    (document as any)._fluentGrid.mediaSM.push({ id: id, dotNetHelper: dotNetHelper });
    (document as any)._fluentGrid.mediaMD.push({ id: id, dotNetHelper: dotNetHelper });
    (document as any)._fluentGrid.mediaLG.push({ id: id, dotNetHelper: dotNetHelper });
    (document as any)._fluentGrid.mediaXL.push({ id: id, dotNetHelper: dotNetHelper });
    (document as any)._fluentGrid.mediaXXL.push({ id: id, dotNetHelper: dotNetHelper });

    // First check
    GetMediaQueries().forEach((mediaQuery) => {
      if (window.matchMedia(mediaQuery.query).matches) {
        dotNetHelper.invokeMethodAsync('FluentGrid_MediaChangedAsync', mediaQuery.id);
      }
    });
  }

  export function FluentGridCleanup(id: string) {
    if ((document as any)._fluentGrid) {
      RemoveItem((document as any)._fluentGrid.mediaXS, id);
      RemoveItem((document as any)._fluentGrid.mediaSM, id);
      RemoveItem((document as any)._fluentGrid.mediaMD, id);
      RemoveItem((document as any)._fluentGrid.mediaLG, id);
      RemoveItem((document as any)._fluentGrid.mediaXL, id);
      RemoveItem((document as any)._fluentGrid.mediaXXL, id);
    }
  }

  // Remove the Array item where item.id is found
  function RemoveItem(array: any, id: string) {
    for (var i = array.length - 1; i >= 0; i--) {
      if (array[i].id === id) {
        array.splice(i, 1);
      }
    }
  }

}
